File: Android/app/src/main/java/com/dkaluta/prosary/models/PrayerRunProgress.kt
package com.dkaluta.prosary.models

import android.content.Context
import java.time.LocalDate
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** The small piece of an interrupted prayer needed to offer Continue on re-entry. The prayer's
 * authored/configured options remain in its favorite or bundle; a run stores only the current
 * step and the language actually selected inside the flow. */
@Serializable
data class PrayerRunProgress(
    val stepIndex: Int,
    val languageCode: String,
    /** ISO local date, intentionally without a timezone: Rosaries expire when the user's local
     * calendar turns over, even if the saved epoch instant would still be yesterday elsewhere. */
    val savedLocalDate: String,
    /** Stable description of the options that produced this run's step sequence. Null keeps
     * old serialized checkpoints decodable, but they are rejected whenever a current signature
     * is supplied rather than risking a resume into a differently configured prayer. */
    val configurationSignature: String? = null,
) {
    fun canResume(
        stepCount: Int,
        today: LocalDate = LocalDate.now(),
        sameLocalDayOnly: Boolean = false,
        expectedConfigurationSignature: String? = null,
    ): Boolean =
        stepIndex in 1 until stepCount &&
            (expectedConfigurationSignature == null || configurationSignature == expectedConfigurationSignature) &&
            (!sameLocalDayOnly || savedLocalDate == today.toString())
}

/** The raw language choice is persisted separately on [PrayerRunProgress]. A custom devotion's
 * resolved form is included because some languages own a structurally different default form;
 * these values describe configuration that can change the generated sequence or visual identity. */
object PrayerRunSignatures {
    fun rosary(options: RosaryOptions): String = listOf(
        "rosary",
        options.mysterySelectionMode.stableValue,
        options.specificMysteryGroup.stableValue,
        options.specificMysteryOrder.toString(),
        options.includeApostlesCreed.flag,
        options.includeOpeningPrayers.flag,
        options.includeOpeningFatimaPrayer.flag,
        options.includeFatimaPrayer.flag,
        options.eternalRestForDeceased.stableValue,
        options.marianAntiphon.stableValue,
        options.includeClosingIntentions.flag,
        options.includeStMichaelPrayer.flag,
        options.includeFinalSignOfCross.flag,
        options.aramaicSignOfCrossForm,
        options.presenterMode.flag,
        options.mysteryImageStyle.stableValue,
    ).joinToString("|")

    fun custom(
        devotionId: String,
        effectiveVariantId: String?,
        dayIndex: Int,
        options: Map<String, String>,
    ): String {
        val optionText = options.toSortedMap().entries.joinToString("|") { (key, value) ->
            "$key=$value"
        }
        return "custom|$devotionId|${effectiveVariantId.orEmpty()}|$dayIndex|$optionText"
    }

    fun jesus(target: JesusPrayerTarget): String = when (target) {
        is JesusPrayerTarget.Count -> "jesus|count|${target.value}"
        JesusPrayerTarget.Unbounded -> "jesus|unbounded"
    }

    private val Enum<*>.stableValue: String
        get() = name.replaceFirstChar { if (it.isUpperCase()) it.lowercaseChar() else it }

    private val Boolean.flag: String get() = if (this) "1" else "0"
}

/** A language normally changes only text. A devotion may instead declare a different default
 * form for a language; carrying a numeric position into that other sequence would open an
 * unrelated prayer. */
object CustomDevotionLanguageSwitch {
    fun indexAfterSwitch(
        currentIndex: Int,
        previousEffectiveVariantId: String?,
        nextEffectiveVariantId: String?,
        nextStepCount: Int,
    ): Int = if (previousEffectiveVariantId == nextEffectiveVariantId) {
        currentIndex.coerceIn(0, (nextStepCount - 1).coerceAtLeast(0))
    } else {
        0
    }
}

/** Stable platform-local run identities, shared in spelling with the other ports. A custom
 * devotion includes form and day because those can produce different step sequences. */
object PrayerRunKeys {
    fun rosary(prayerId: String): String = "rosary:$prayerId"

    fun custom(devotionId: String, variantId: String?, dayIndex: Int): String =
        "custom:$devotionId:${variantId.orEmpty()}:$dayIndex"

    fun jesus(prayerId: String?, target: JesusPrayerTarget): String =
        "jesus:${prayerId ?: target.toRouteValue()}"
}

/** Platform-local transient run storage. Keys are stable prayer identities, allowing multiple
 * saved Rosaries to keep their own place while ad-hoc Jesus Prayer sessions use their target. */
object PrayerRunProgressStore {
    private const val PREFS = "prayer_run_progress"
    private const val KEY = "runs"
    private val json = Json { ignoreUnknownKeys = true }

    private fun prefs(context: Context) = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun all(context: Context): Map<String, PrayerRunProgress> {
        val raw = prefs(context).getString(KEY, null) ?: return emptyMap()
        return runCatching { json.decodeFromString<Map<String, PrayerRunProgress>>(raw) }
            .getOrDefault(emptyMap())
    }

    private fun saveAll(context: Context, runs: Map<String, PrayerRunProgress>) {
        prefs(context).edit().putString(KEY, json.encodeToString(runs)).apply()
    }

    fun progress(context: Context, runKey: String): PrayerRunProgress? = all(context)[runKey]

    fun save(
        context: Context,
        runKey: String,
        stepIndex: Int,
        languageCode: String,
        configurationSignature: String? = null,
        today: LocalDate = LocalDate.now(),
    ) {
        if (stepIndex <= 0) {
            clear(context, runKey)
            return
        }
        saveAll(
            context,
            all(context) + (runKey to PrayerRunProgress(
                stepIndex = stepIndex,
                languageCode = languageCode,
                savedLocalDate = today.toString(),
                configurationSignature = configurationSignature,
            )),
        )
    }

    fun clear(context: Context, runKey: String) {
        val remaining = all(context) - runKey
        if (remaining.isEmpty()) {
            prefs(context).edit().remove(KEY).apply()
        } else {
            saveAll(context, remaining)
        }
    }
}

File: iOS/Prosary/Models/PrayerRunProgress.swift
//
//  PrayerRunProgress.swift
//  Prosary
//
//  A deliberately small, device-local bookmark for an interrupted prayer. Authored content
//  and configuration remain in the prayer preset/bundle; this stores only the current step,
//  the language actually used inside the flow, and the local civil day needed by the Rosary's
//  same-day continuation rule.
//

import Foundation

struct PrayerRunProgress: Codable, Equatable {
  /// A stable description of the sequence-producing options. An old bookmark must not land
  /// on an unrelated prayer after its preset, variant, day, or repeat target changes.
  let configurationSignature: String?
  let stepIndex: Int
  let languageCode: String
  /// ISO local date without a timezone. A Rosary expires when the user's local calendar turns
  /// over, irrespective of what date that instant represents in another timezone.
  let savedLocalDate: String

  init(
    configurationSignature: String? = nil,
    stepIndex: Int,
    languageCode: String,
    savedLocalDate: String
  ) {
    self.configurationSignature = configurationSignature
    self.stepIndex = stepIndex
    self.languageCode = languageCode
    self.savedLocalDate = savedLocalDate
  }

  func canResume(
    stepCount: Int,
    today: Date = Date(),
    calendar: Calendar = .current,
    sameLocalDayOnly: Bool = false,
    expectedConfigurationSignature: String? = nil
  ) -> Bool {
    stepIndex > 0 && stepIndex < stepCount
      && (expectedConfigurationSignature == nil
        || configurationSignature == expectedConfigurationSignature)
      && (!sameLocalDayOnly || savedLocalDate == Self.localDateString(for: today, calendar: calendar))
  }

  static func localDateString(for date: Date, calendar: Calendar = .current) -> String {
    let components = calendar.dateComponents([.year, .month, .day], from: date)
    return String(
      format: "%04d-%02d-%02d",
      locale: Locale(identifier: "en_US_POSIX"),
      components.year ?? 0,
      components.month ?? 0,
      components.day ?? 0)
  }
}

/// Platform-local transient storage. Keys are stable prayer identities (`rosary:<uuid>`,
/// `custom:<bundle>:<variant>:<day>`, `jesus:<uuid-or-target>`), allowing multiple Rosary presets
/// to retain independent places without applying a bookmark to another form/day of a devotion.
struct PrayerRunProgressStore {
  static let defaultsKey = "prayerRunProgress"

  private let defaults: UserDefaults
  private let defaultsKey: String
  private let namespace: String?

  init(defaults: UserDefaults = .standard, defaultsKey: String = Self.defaultsKey, namespace: String? = nil) {
    self.defaults = defaults
    self.defaultsKey = defaultsKey
    self.namespace = namespace
  }

  func progress(for runKey: String) -> PrayerRunProgress? {
    var runs = all()
    let key = storageKey(runKey)
    if let ownProgress = runs[key] { return ownProgress }
    // Existing installs and newly opened windows resume the most recently saved session.
    // Claim a copy once; later sibling-window saves must not move this window's bookmark.
    guard namespace != nil, let latestProgress = runs[runKey] else { return nil }
    runs[key] = latestProgress
    save(runs)
    return latestProgress
  }

  func save(
    runKey: String,
    stepIndex: Int,
    languageCode: String,
    configurationSignature: String? = nil,
    today: Date = Date(),
    calendar: Calendar = .current
  ) {
    guard stepIndex > 0 else {
      clear(runKey: runKey)
      return
    }
    var runs = all()
    let progress = PrayerRunProgress(
      configurationSignature: configurationSignature,
      stepIndex: stepIndex,
      languageCode: languageCode,
      savedLocalDate: PrayerRunProgress.localDateString(for: today, calendar: calendar))
    runs[storageKey(runKey)] = progress
    // Retain a shared continuation for an explicitly closed window, whose scene namespace
    // won't return when the person later chooses File → New Window.
    if namespace != nil { runs[runKey] = progress }
    save(runs)
  }

  func clear(runKey: String) {
    var runs = all()
    let removedProgress = runs.removeValue(forKey: storageKey(runKey))
    // Finishing an older window must not erase the continuation most recently saved by a
    // sibling. The shared copy belongs to this window only while its checkpoint still matches.
    if namespace != nil, let removedProgress, runs[runKey] == removedProgress {
      runs.removeValue(forKey: runKey)
    }
    save(runs)
  }

  /// A restored Mac scene retains this namespace; sibling windows never share a bookmark.
  /// Mobile keeps the existing prayer-identity keys.
  private func storageKey(_ runKey: String) -> String {
    namespace.map { "window:\($0):\(runKey)" } ?? runKey
  }

  private func all() -> [String: PrayerRunProgress] {
    guard let data = defaults.data(forKey: defaultsKey) else { return [:] }
    return (try? JSONDecoder().decode([String: PrayerRunProgress].self, from: data)) ?? [:]
  }

  private func save(_ runs: [String: PrayerRunProgress]) {
    guard !runs.isEmpty else {
      defaults.removeObject(forKey: defaultsKey)
      return
    }
    guard let data = try? JSONEncoder().encode(runs) else { return }
    defaults.set(data, forKey: defaultsKey)
  }
}

enum PrayerRunKey {
  static func rosary(_ prayer: Prayer) -> String { "rosary:\(prayer.id.uuidString)" }
  static func custom(_ devotionId: String, variantId: String?, dayIndex: Int) -> String {
    "custom:\(devotionId):\(variantId ?? ""):\(dayIndex)"
  }

  static func jesus(_ prayer: Prayer?, target: JesusPrayerTarget) -> String {
    if let prayer { return "jesus:\(prayer.id.uuidString)" }
    switch target {
    case .count(let count): return "jesus:\(count)"
    case .unbounded: return "jesus:unbounded"
    }
  }
}

/// The raw language choice is persisted separately in `PrayerRunProgress`. A custom devotion's
/// resolved form is included because some languages own a structurally different default form;
/// everything here can change the generated sequence or its visual identity.
enum PrayerRunSignature {
  static func rosary(_ options: RosaryOptions) -> String {
    var fields = [
      "rosary",
      options.mysterySelectionMode.rawValue,
      options.specificMysteryGroup.rawValue,
      String(options.specificMysteryOrder),
      flag(options.includeApostlesCreed),
      flag(options.includeOpeningPrayers),
      flag(options.includeOpeningFatimaPrayer),
      flag(options.includeFatimaPrayer),
      options.eternalRestForDeceased.rawValue,
      options.marianAntiphon.rawValue,
      flag(options.effectiveClosingIntentions),
      flag(options.includeStMichaelPrayer),
      flag(options.includeFinalSignOfCross),
      options.aramaicSignOfCrossForm,
      flag(options.presenterMode),
      options.mysteryImageStyle.rawValue,
    ]
    // Preserve whole-group bookmarks; former partial groups now have a different sequence.
    if options.effectiveClosingIntentions {
      fields.append("closing-v2:1,1,1")
    }
    if options.includeOpeningPrayers && options.includeOpeningFatimaPrayer {
      fields.append("opening-fatima-v2")
    }
    return fields.joined(separator: "|")
  }

  static func custom(
    _ devotionId: String,
    effectiveVariantId: String?,
    dayIndex: Int,
    options: [String: String]
  ) -> String {
    let options = RosaryOptions.normalizedCustomOptions(options, bundleId: devotionId)
    let optionText = options.keys.sorted().map { "\($0)=\(options[$0] ?? "")" }.joined(separator: "|")
    var signature = "custom|\(devotionId)|\(effectiveVariantId ?? "")|\(dayIndex)|\(optionText)"
    if devotionId == "rosary", options["closingIntentions"] == "true" {
      signature += "|closing-v2:1,1,1"
    }
    if devotionId == "rosary", options["openingPrayers"] != "false", options["openingFatimaPrayer"] == "true" {
      signature += "|opening-fatima-v2"
    }
    return signature
  }

  static func jesus(_ target: JesusPrayerTarget) -> String {
    switch target {
    case .count(let count): return "jesus|count|\(count)"
    case .unbounded: return "jesus|unbounded"
    }
  }

  private static func flag(_ value: Bool) -> String { value ? "1" : "0" }
}

/// Pure navigation targets for the Rosary toolbar. Moving between mysteries changes only the
/// current step index, leaving the ordinary bead-by-bead Back/Next controls and BeadLayout intact.
enum RosaryMysteryNavigation {
  static func announcementIndices(in steps: [RosaryStep]) -> [Int] {
    var seen: Set<Int> = []
    return steps.indices.filter { index in
      guard let decadeIndex = steps[index].decadeIndex else { return false }
      return seen.insert(decadeIndex).inserted
    }
  }

  static func previousIndex(in steps: [RosaryStep], from currentIndex: Int) -> Int? {
    let starts = announcementIndices(in: steps)
    if let currentDecade = steps.indices.contains(currentIndex) ? steps[currentIndex].decadeIndex : nil {
      guard currentDecade > 0 else { return nil }
      return starts.first { steps[$0].decadeIndex == currentDecade - 1 }
    }
    return starts.last { $0 < currentIndex }
  }

  static func nextIndex(in steps: [RosaryStep], from currentIndex: Int) -> Int? {
    let starts = announcementIndices(in: steps)
    if let currentDecade = steps.indices.contains(currentIndex) ? steps[currentIndex].decadeIndex : nil {
      return starts.first { steps[$0].decadeIndex == currentDecade + 1 }
    }
    return starts.first { $0 > currentIndex }
  }
}

/// A language normally changes only text. Some devotions, however, declare a different default
/// form for a language; carrying a numeric position into that other sequence would open an
/// unrelated prayer.
enum CustomDevotionLanguageSwitch {
  static func indexAfterSwitch(
    currentIndex: Int,
    previousEffectiveVariantId: String?,
    nextEffectiveVariantId: String?,
    nextStepCount: Int
  ) -> Int {
    guard previousEffectiveVariantId == nextEffectiveVariantId else { return 0 }
    return min(currentIndex, max(nextStepCount - 1, 0))
  }
}
